### PR TITLE
initial version of realtimer

### PR DESCRIPTION
**Pull request by mat-hek**
todo:
- [x] tests
- [x] documentation
- [ ] skipping late packets -> #3 
- [ ] consider using pull output (?) -> #4

### CODE REVIEW

File: mix.exs [hunk 1]
```diff
@@ -1,24 +1,24 @@
-defmodule Membrane.Template.Mixfile do
+defmodule Membrane.Realtimer.Plugin.Mixfile do
   use Mix.Project
 
   @version "0.1.0"
   @github_url "https://github.com/membraneframework/membrane_template_plugin"
```
Comment (mickel8) [inline mix.exs]:
```suggestion
  @github_url "https://github.com/membraneframework/membrane_realtimer_plugin"
  ```

File: mix.exs [hunk 2]
```diff
@@ -1,24 +1,24 @@
...
+      app: :membrane_realtimer_plugin,
       version: @version,
       elixir: "~> 1.9",
       elixirc_paths: elixirc_paths(Mix.env()),
       start_permanent: Mix.env() == :prod,
       deps: deps(),
 
       # hex
-      description: "Template Plugin for Membrane Multimedia Framework",
+      description: "Membrane element limiting playback speed to realtime",
```
Comment (mickel8) [inline mix.exs]:
```suggestion
      description: "Membrane plugin for limiting playback speed to realtime",
```

File: README.md [hunk 3]
```diff
@@ -1,34 +1,71 @@
...
-[![Hex.pm](https://img.shields.io/hexpm/v/membrane_template_plugin.svg)](https://hex.pm/packages/membrane_template_plugin)
-[![CircleCI](https://circleci.com/gh/membraneframework/membrane_template_plugin.svg?style=svg)](https://circleci.com/gh/membraneframework/membrane_template_plugin)
+[![Hex.pm](https://img.shields.io/hexpm/v/membrane_realtimer_plugin.svg)](https://hex.pm/packages/membrane_realtimer_plugin)
+[![API Docs](https://img.shields.io/badge/api-docs-yellow.svg?style=flat)](https://hexdocs.pm/membrane_realtimer_plugin/)
+[![CircleCI](https://circleci.com/gh/membraneframework/membrane_realtimer_plugin.svg?style=svg)](https://circleci.com/gh/membraneframework/membrane_realtimer_plugin)
 
-This repository contains template a template for new elements.
-
-Check out different branches for other flavours of template
+Membrane element limiting playback speed to realtime, according to buffers' timestamps.
```
Comment (mickel8) [inline README.md]:
```suggestion
Membrane plugin for limiting playback speed to realtime, according to buffers' timestamps.
```

File: README.md [hunk 4]
```diff
@@ -1,34 +1,71 @@
-# Membrane Multimedia Framework: Template Plugin
+# Membrane Realtimer Plugin
```
Comment (mickel8) [inline README.md]:
```suggestion
# Membrane Realtimer plugin
```
In all repos we have `Membrane Xyz plugin`. `Membrane Xyz Plugin` looks better imo but let's follow the convention.

File: test/membrane_realtimer_test.exs [hunk 5]
```diff
@@ -0,0 +1,32 @@
+defmodule Membrane.FunnelTest do
```
Comment (mickel8) [inline test/membrane_realtimer_test.exs]:
```suggestion
defmodule Membrane.RealtimerTest do
```

File: test/membrane_realtimer_test.exs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  use ExUnit.Case
+
+  import Membrane.Testing.Assertions
+
+  alias Membrane.{Buffer, Testing, Time, Realtimer}
+
+  test "Collects multiple inputs" do
```
Comment (mickel8) [inline test/membrane_realtimer_test.exs]:
```suggestion
  test "Limits playback speed to realtime" do
```

File: lib/membrane/realtimer.ex [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+  If buffers come in slower than realtime, they're sent as they come in.
+  """
+  use Membrane.Filter
+
+  def_input_pad :input, caps: :any, demand_unit: :buffers
+  def_output_pad :output, caps: :any, mode: :push
+
+  @impl true
+  def handle_init(_opts) do
+    {:ok, %{timestamp: 0, tick_actions: []}}
```
Comment (mickel8) [inline lib/membrane/realtimer.ex]: `timestamp` -> `last_buffer_timestamp` ?

File: lib/membrane/realtimer.ex [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+  end
+
+  @impl true
+  def handle_prepared_to_playing(_ctx, state) do
+    {{:ok, start_timer: {:timer, :no_interval}, demand: {:input, 1}}, state}
+  end
+
+  @impl true
+  def handle_process(:input, buffer, _ctx, state) do
```
Comment (mickel8) [inline lib/membrane/realtimer.ex:23]: Just to be sure. In ech `tick_actions` list we will always have only one buffer to convey?

File: lib/membrane/realtimer.ex [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    actions =
+      [timer_interval: {:timer, :no_interval}] ++
+        Enum.reverse(state.tick_actions) ++ [demand: {:input, 1}]
+
+    {{:ok, actions}, %{state | tick_actions: []}}
+  end
+
+  @impl true
+  def handle_playing_to_prepared(_ctx, state) do
+    {{:ok, stop_timer: :timer}, state}
```
Comment (mickel8) [inline lib/membrane/realtimer.ex]: Shouldn't we reset `timestamp` here?

**Review by mickel8 — CHANGES_REQUESTED** [review]

File: lib/membrane/realtimer.ex (hunk 8 above)
Comment (mat-hek) [reply lib/membrane/realtimer.ex:23]: yes, but events and caps may also be there

**Review by mickel8 — APPROVED** [review]

**Pull request merged**